File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BitManipulation/29_DivideTwoIntegers.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


c.. Solution o..
    # According to:
    # https://leetcode.com/discuss/38997/detailed-explained-8ms-c-solution
    # Key concept:
    # division simply requires us to find how many times we can subtract the
    # divisor from the the dividend without making the dividend negative.

    ___ divide  dividend, divisor
        __ divisor __ 0:
            r_ -1

        # Make sure it's positive or negative
        positive = (dividend < 0) is (divisor < 0)
        dividend, divisor = abs(dividend), abs(divisor)
        answer = 0

        _____ dividend >= divisor:
            multiple, temp = 1, divisor
            _____ dividend >= (temp << 1
                multiple <<= 1
                temp <<= 1
            dividend -= temp
            answer += multiple

        __ n.. positive:
            answer = -answer

        __ (answer > 2147483647) or (answer < -2147483648
            r_ 2147483647
        r_ answer

"""
0
1
12
3
125
-4
1
-1
"""

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BitManipulation/29_DivideTwoIntegers.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-06 22:39:41
 */


c.. Solution {
public:
    // Refer to:
    // https://leetcode.com/discuss/38997/detailed-explained-8ms-c-solution
    int divide(int dividend, int divisor) {
        __(!divisor || (dividend__INT_MIN && divisor__-1))
            r_ INT_MAX;

        bool positive = ((dividend < 0) __ (divisor < 0));
        long long dvd = labs(dividend);
        long long dvs = labs(divisor);
        int ans = 0;
        _____(dvd >= dvs){
            long long temp=dvs, multiple=1;
            _____(dvd >= (temp<<1)){
                temp <<= 1;
                multiple <<= 1;
            }
            ans += multiple;
            dvd -= temp;
        }

        r_ positive ? ans:-ans;
    }
};

/*
0
1
12
3
125
-4
1
-1
2147483647
1
*/
